dask/array/creation.py

```diff
@@ -496,7 +496,7 @@ def diag(v):
             if i == j:
                 dsk[key] = (np.diag, blocks[i])
             else:
-                dsk[key] = (np.zeros, (m, n))
+                dsk[key] = (partial(np.zeros_like, shape=(m,n)), v._meta)
```

Comment:
My guess is that we may not be able to make this work short term

Reply:
Unfortunately, that's true. I will probably split this PR into two different ones, the first one to do all the things we can support right away, the other for unreleased NumPy features.